core/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java

```diff
@@ -899,6 +903,232 @@ public void testDataFileCorruptionDuringRestore() throws Exception {
+                assertThat(unassignedInfo.getReason(), equalTo(UnassignedInfo.Reason.ALLOCATION_FAILED));
+                assertThat(unassignedInfo.getNumFailedAllocations(), equalTo(maxRetries));
+                assertThat(unassignedInfo.getLastAllocationStatus(), equalTo(UnassignedInfo.AllocationStatus.DECIDERS_NO));
+            }
+        });
+
+        ClusterStateResponse clusterStateResponse = client().admin().cluster().prepareState().setCustoms(true).setRoutingTable(true).get();
+
+        // check that there is no restore in progress
+        RestoreInProgress restoreInProgress = clusterStateResponse.getState().custom(RestoreInProgress.TYPE);
+        assertNotNull("RestoreInProgress must be not null", restoreInProgress);
+        assertThat("RestoreInProgress must be empty", restoreInProgress.entries(), hasSize(0));
+
+        // check that the shards have been created but are not assigned
+        List<ShardRouting> shards = clusterStateResponse.getState().getRoutingTable().allShards(indexName);
+        assertThat(shards, hasSize(numShards.totalNumShards));
+        shards.forEach(shard -> assertTrue("Shard " + shard.shardId() + " must be unassigned", shard.unassigned()));
+
+        // update the test repository in order to make it work
+        assertAcked(client().admin().cluster().preparePutRepository("test-repo")
+                                                .setType("fs")
+                                                .setSettings(Settings.builder().put("location", repositoryLocation)));
+
+        final boolean useReroute = randomBoolean();
+        if (useReroute) {
+            // Option 1: use the Reroute API to retry the failed shards
+            assertAcked(client().admin().cluster().prepareReroute().setRetryFailed(true));
+
+        } else {
+            // Option 2: delete the index and restore again
+            assertAcked(client().admin().indices().prepareDelete(indexName));
+            restoreResponse = client().admin().cluster().prepareRestoreSnapshot("test-repo", "test-snap")
+                                                        .setMasterNodeTimeout("30s")
+                                                        .setWaitForCompletion(true)
+                                                        .get();
+            assertThat(restoreResponse.getRestoreInfo().totalShards(), equalTo(numShards.numPrimaries));
+            assertThat(restoreResponse.getRestoreInfo().successfulShards(), equalTo(numShards.numPrimaries));
+        }
+
+        // Wait for the shards to be assigned
+        waitForRelocation();
+        refresh(indexName);
+
+        assertThat(client().prepareSearch(indexName).setSize(0).get().getHits().getTotalHits(), equalTo((long) nbDocs));
+    }
+
+    /**
+     * Test that restoring an index with shard allocation filtering settings that prevents
+     * its allocation does not hang indefinitely.
+     */
+    public void testUnrestorableIndexDuringRestore() throws Exception {
```

Comment:
I think it's possible to share most of the code with the previous test by calling a generic method with two parameters:
- restoreIndexSettings (which would set maxRetries in first case and filters in second case)
- fixupAction (the action to run to fix the issue)

Reply:
I tried to do something like that, please let me know what you think.